qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def symmetric(self):
+    return self._symmetric
+
+  @symmetric.setter
+  def symmetric(self, symmetric):
+    self._symmetric = symmetric
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def keep_negative(self):
+    return self._keep_negative
+
+  @keep_negative.setter
+  def keep_negative(self, keep_negative):
+    self._keep_negative = keep_negative
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def qnoise_factor(self):
+    return self._qnoise_factor
+
+  @qnoise_factor.setter
+  def qnoise_factor(self, qnoise_factor):
+    self._qnoise_factor = qnoise_factor
+
+  @property
+  def use_stochastic_rounding(self):
+    return self._use_stochastic_rounding
+
+  @use_stochastic_rounding.setter
+  def use_stochastic_rounding(self, use_stochastic_rounding):
+    self._use_stochastic_rounding = use_stochastic_rounding
+
+  @property
+  def scale_axis(self):
+    return self._scale_axis
+
+  @scale_axis.setter
+  def scale_axis(self, scale_axis):
+    self._scale_axis = scale_axis
+
+  @property
+  def alpha(self):
+    return self._alpha
+
+  @alpha.setter
+  def alpha(self, alpha):
+    """Set alpha and check if it isvalid."""
+
+    self._alpha = alpha
+
+    # Error checking
+    if isinstance(alpha, six.string_types):
+      # Check the quantizer has been given a valid alpha string
+      if not alpha in self.ALPHA_OPTIONS:
+        raise ValueError(
+            f"Invalid alpha '{alpha}' for auto alpha computation. "
+            f"Must be one of {self.ALPHA_OPTIONS}")
+    elif alpha is not None: # alpha is a tensor
+      try:
+        # any allowable array type can be cast as a numpy array
+        np.array(alpha)
+      except TypeError:
+        raise TypeError(
+            f"alpha must be, a string, an array, or None, not {type(alpha)}")
+
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def scale(self):
+    return self.quantization_scale / self.data_type_scale
+  
+  @property
+  def quantization_scale(self):
+    return self._quantization_scale
+  
+  @quantization_scale.setter
+  def quantization_scale(self, quantization_scale):
+    self._quantization_scale = quantization_scale
+    
+  @property
+  def data_type_scale(self):
+    """Quantization scale for the data type"""
+    # integer is sometimes cast as int32, so cast to float32 to avoid errors
+    integer = tf.cast(self.integer, tf.float32)
+    return K.pow(2.0, integer - self.bits + self.keep_negative)
+
+  @property
+  def auto_alpha(self):
+    """Returns true if using a data-dependent alpha"""
+
+    return isinstance(self.alpha, six.string_types)
+
+  @property
+  def use_sign_function(self):
+    """Return true if using sign function for quantization"""
+
+    return (self.bits == 1.0) and self.keep_negative
+
+  @property
+  def clip_bounds(self):
```

Comment:
Why is this a property instead of a normal function called "calculate_clip_bounds"?

Reply:
They're essentially equivalent. Since it's only data that's determinable from initialization parameters, I thought it made sense to treat it like an attribute

Reply:
Agreed